substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/ThreadMXUtils.java

```diff
@@ -0,0 +1,176 @@
+import com.oracle.svm.core.thread.JavaThreads.JMXMonitoring;
+import com.oracle.svm.core.thread.PlatformThreads;
+import org.graalvm.collections.EconomicSet;
+
+import java.lang.management.ThreadInfo;
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.concurrent.locks.AbstractOwnableSynchronizer;
+import java.util.concurrent.locks.LockSupport;
+
+/**
+ * Utils to support {@link SubstrateThreadMXBean} for providing threading information for JMX
+ * support. Include the {@link ThreadInfo} constructing utils, and deadlock detection utils.
+ */
+public class ThreadMXUtils {
+
+    public static class ThreadInfoConstructionUtils {
+
+        private static StackTraceElement[] getStackTrace(Thread thread, int maxDepth) {
+            StackTraceElement[] stackTrace = thread.getStackTrace();
+            return maxDepth == -1 || maxDepth >= stackTrace.length ? stackTrace : Arrays.copyOfRange(stackTrace, 0, maxDepth);
+        }
+
+        private record Blocker(Object blockObject, Thread ownerThread) {
+        }
+
+        private static Blocker getBlockerInfo(Thread thread) {
+            Object blocker = LockSupport.getBlocker(thread);
+
+            if (blocker instanceof JavaMonitor javaMonitor) {
+                return new Blocker(
+                                javaMonitor.getBlockedObject(),
+                                SubstrateThreadMXBean.getThreadById(javaMonitor.getOwnerThreadId()));
+            } else if (blocker instanceof AbstractOwnableSynchronizer synchronizer) {
+                return new Blocker(synchronizer,
+                                SubstrateUtil.cast(synchronizer, Target_java_util_concurrent_locks_AbstractOwnableSynchronizer.class)
+                                                .getExclusiveOwnerThread());
+            }
+            return new Blocker(blocker, null);
+        }
+
+        private static Object[] getLockedSynchronizers(Thread thread) {
+            EconomicSet<AbstractOwnableSynchronizer> locks = JMXMonitoring.getThreadLocks(thread);
+            if (locks != null) {
+                synchronized (locks) {
+                    Object[] lockObjects = new Object[locks.size()];
+                    Iterator<AbstractOwnableSynchronizer> iterator = locks.iterator();
+                    for (int i = 0; iterator.hasNext(); i++) {
+                        lockObjects[i] = iterator.next();
+                    }
+                    return lockObjects;
+                }
+            }
+            return new Object[0];
+        }
+
+        private record LockedMonitors(Object[] monitorObjects, int[] monitorDepths) {
+        }
+
+        private static LockedMonitors getLockedMonitors(Thread thread, int stacktraceLength) {
+            ArrayDeque<JMXMonitoring.MonitorInfo> monitors = JMXMonitoring.getThreadMonitors(thread);
+            if (monitors != null) {
+                synchronized (monitors) {
```

Comment:
I believe there is still a race here.  I don't think it's enough to only synchronize here when we're reading the monitors. We also need to synchronize when we modify the stack of monitors in `removeThreadMonitor`  and `addThreadMonitor`.  For example, if `monitors` is modified in between when `monitorObjects` and `monitorDepths` is assigned, `monitorObjects` and `monitorDepths` won't be the same size.

Reply:
Hi @roberttoyonaga, I see.
So, only the thread that owns the monitors is modifying the monitors collection. The other thread is reading it here. I have made a change to get an unmodifiable copy of the monitors list, so then I can use it without worrying about changes in the original monitors collection. I think I shouldn't block the writes when I'm reading the collection.

Reply:
Thank you. I think this idea of getting a "snapshot" of the list is good. However, there is still the possibility that the list is modified while `List.copyOf()` is in-progress (race during construction).  I'm not sure how bad this could be because it depends on the implementation of  `List.copyOf()` , `ArrayDeque.push` `ArrayDeque.pop`